src/appsupport.c

```diff
@@ -358,11 +358,24 @@ static void appRenameItem(item_list_t *itemList, int id, char *newName)
 static void appLaunchItem(item_list_t *itemList, int id, config_set_t *configSet)
 {
     int fd;
-    const char *filename;
+    char filename[256];
     const char *argv1;
 
     // Retrieve configuration set by appGetConfig()
-    configGetStr(configSet, CONFIG_ITEM_STARTUP, &filename);
+    configGetStrCopy(configSet, CONFIG_ITEM_STARTUP, filename, sizeof(filename));
+
+    // Legacy apps state a full filepath eg mass0.. but we don't know which device number the app in question is mounted to so we need to check where file exists before trying to launch it.
+    // we either find it and break keeping the device number or its not found on any and fall through to an eventual error message.
+    if (!strncmp("mass", filename, 4)) {
+        for (int i = 0; i < BDM_MODE4; i++) {
```

Comment:
What happens when I have an old app on `mass0`, and a new app (same name) on 'mass1'. This code will always load the old app from `mass0` I think? Even if it's not legacy apps?

I would expect the correct path to be put into CONFIG_ITEM_STARTUP when loading. Becouse after the wrong paths have been loaded, it's difficult to reverse trial-and-error figure out what it should have been.

Reply:
> What happens when I have an old app on mass0, and a new app (same name) on 'mass1'. This code will always load the old app from mass0 I think? Even if it's not legacy apps?

Yea that would happen, multiple issues here due to the legacy apps system I'm not sure the best way around it. The main issue is the user inputs a path in their cfg `LEGACYSMS=`mass0:/SMS.ELF`` this is the example I tested with.. we have no idea which device they're trying to access at any point in the program as device numbers are all given dependant on which is connected when. So when reading `mass0:/SMS.ELF` this into CONFIG_ITEM_STARTUP we still don't know which device they want.

Reply:
I think FMCB handles this using `mass?:`, and we can probably do the same? If a user configs `mass0:` then I think it's good practice to only try loading from `mass0:`. If autodetection is wanted by the user then `mass?:` can trigger the above autodetection loop.
The default configs should probably become `mass?:` in this case, for easy plug and play. If a user wants to force loading from mass1, then he can change it to `mass1:`

Reply:
FMCB/FHDB/Funtuna does just use the first thing where it detects a USB-Storage-device + the folder(s) and ELF. It doesn't really have any "multi storage device support".

The drivers (IRXs) do, but the Payload/"loader" does not.

Reply:
That's how FMCB itself gets loaded, but I means after it's loaded, then a config (configured using FMCB configurator) loads an elf depending on what button is pressed.
That ELF can be loaded from `mass?:`.

Reply:
Oh? FMCB can (already) define from which port?

If it can, then SP193 must have implemented in the 1.9x series, AFAIK.

Reply:
I updated the previous post as well with additional info.

Reply:
> I think FMCB handles this using `mass?:`, and we can probably do the same? If a user configs `mass0:` then I think it's good practice to only try loading from `mass0:`. If autodetection is wanted by the user then `mass?:` can trigger the above autodetection loop. The default configs should probably become `mass?:` in this case, for easy plug and play. If a user wants to force loading from mass1, then he can change it to `mass1:`

I've adjusted it to only run the check if value is mass? https://github.com/ps2homebrew/Open-PS2-Loader/pull/1289/commits/00ce8f7773ddd40c661ce0a43097390ff6c1d667 because of hot plugging devices, this check should be performed at launch rather than when the apps list is populated right?